expression/grouping_sets.go

```diff
@@ -0,0 +1,375 @@
+
+// Merge function will explore the internal grouping expressions and try to find the minimum grouping sets. (prefix merging)
+func (gss GroupingSets) Merge() GroupingSets {
+	// for now, there is precondition that all grouping expressions are columns.
+	// for example: (a,b,c) and (a,b) and (a) will be merged as a one.
+	// Eg:
+	// before merging, there are 4 grouping sets.
+	// GroupingSets:
+	//     [
+	//         [[a,b,c],]             // every set including a grouping Expressions for initial.
+	//         [[a,b],]
+	//         [[a],]
+	//         [[e],]
+	//     ]
+	//
+	// after merging, there is only 2 grouping set.
+	// GroupingSets:
+	//     [
+	//         [[a],[a,b],[a,b,c],]   // one set including 3 grouping Expressions after merging if possible.
+	//         [[e],]
+	//     ]
+	//
+	// care about the prefix order, which should be taken as following the group layout expanding rule. (simple way)
+	// [[a],[b,a],[c,b,a],] is also conforming the rule, gradually including one/more column(s) inside for one time.
+
+	newGroupingSets := make(GroupingSets, 0, len(gss))
+	for _, oneGroupingSet := range gss {
+		for _, oneGroupingExpr := range oneGroupingSet {
+			if len(newGroupingSets) == 0 {
+				// means there is nothing in new grouping sets, adding current one anyway.
+				newGroupingSets = append(newGroupingSets, newGroupingSet(oneGroupingExpr))
+				continue
+			}
+			newGroupingSets = newGroupingSets.mergeOne(oneGroupingExpr)
+		}
+	}
+	return newGroupingSets
+}
+
+func (gss GroupingSets) mergeOne(targetOne GroupingExprs) GroupingSets {
+	// for every existing grouping set, check the grouping-exprs inside and whether the current grouping-exprs is
+	// super-set of it or sub-set of it, adding current one to the correct position of grouping-exprs slice.
+	//
+	//                  [[a,b]]
+	//                     |
+	//               /  offset 0
+	//              |
+	//             [b]    when adding [b] grouping expr here, since it's a sub-set of current [a,b] with offset 0, take the offset 0.
+	//              |
+	//     insert with offset 0, and the other elements move right.
+	//
+	//                 [[b],   [a,b]]
+	//             offset 0       1
+	//                                \
+	//                                [a,b,c,d]
+	//        when adding [a,b,c,d] grouping expr here, since it's a super-set of current [a,b] with offset 1, take the offset as 1+1.
+	//
+	//     result grouping set:  [[b], [a,b], [a,b,c,d]]， expanding with step with two or more columns is acceptable and reasonable.
+	//                             ｜    ｜       ｜
+	//                              +----+-------+       every previous one is the subset of the latter one.
+	//
+	for i, oneNewGroupingSet := range gss {
+		// for every group set，try to find its position to insert if possible，otherwise create a new grouping set.
+		for j := len(oneNewGroupingSet) - 1; j >= 0; j-- {
+			cur := oneNewGroupingSet[j]
+			if targetOne.SubSetOf(cur) {
```

Comment:
If the targetOne is super set of cur and j=0, then it will create the new GroupingSet at the line:126.
For example:
```
targetOne is [a,b]
gss is [
   [[a],]
]
```
The result will be 
```
gss [
   [[a]]
   [[a,b]]
]
```

The result is not expected by your comment. Are you sure this logic is correctly ?

Reply:
great catch, fixed~

Reply:
Fixed, LGTM